airflow/executors/base_executor.py

```diff
@@ -232,14 +255,14 @@ def get_event_buffer(self, dag_ids=None) -> Dict[TaskInstanceKeyType, Optional[s
 
     def execute_async(self,
                       key: TaskInstanceKeyType,
-                      command: CommandType,
+                      deferred_run: LocalTaskJobDeferredRun,
                       queue: Optional[str] = None,
```

Comment:
nit: the number of concepts in the name for `LocalTaskJobDeferredRun` begs so many questions about what the hec that thing is and how the whole OOP design is evolving such that complex combinations of so many concepts is necessary in this class name.  e.g. why is it not simple `DeferredTask` or `LocalTask` or `QueuedTask` (and `Queued` implies `DeferredRun` and what's the difference between `Task` and `Job` and why does this one class capture both concepts at the same time?

Reply:
I agree with @dazza-codes here. At first glance, the name LocalTaskJobDeferredRun has a LOT going on it and is kind of hard to understand for people not well acquainted with the code.

Reply:
QueuedTask

> maybe execute_async is better named queue_job?

I did not want changes in this PR that are not necessary. I will change the name of this method in a moment. 

Reply:
@mik-laj Here's an idea: do we actually need a whole new class, or could we instead do something like `executor.execute_async(task_instance)`?

Since the job of the executor is to run execute tasks, not run arbitrary commands, why not just pass it the entire TI object and let the executor work out what it needs to do. Thinking about it now passing down a command to `airflow run `/`airflow tasks run` instead of the TI itself feels like a very leaky abstraction.

Reply:
@ashb We need to store startup configuration information. TaskInstance and SimpleTaskInstance do not store type information. These objects are created in a completely different place and I would not like to mix classes. I like the principle of one responsibility. 

What do you think about "QueueTaskInstance"? It sounds strange, but it makes sense in my opinion.
TaskInstance -> Database entity.
SimpleTaskInstance -> Picklable representation of TaskInstance. This It is used for communication between DagProcecssorAgent and Scheduler. 
QueueTaskInstance -> TaskInstancec identifier with run configuration. This is used for "communication" between Executor and LocalTaskJob.

I think the RawTaskDeferredRun class can be combined with TaskRunner. We do not need to have a class, because we immediately replace the object with a command. Does it make sense to you?

CC: @nuclearpinguin

Reply:
Oh right all the `ignore_*` etc. Gotcha. I somehow managed to forget all of those.

I think we should avoid "queue" here as Airflow already has a concept of queues, and this is not what we are dealing with at this level.

I think Celery calls this a "Request" http://docs.celeryproject.org/en/latest/reference/celery.worker.request.html - how about TaskExecutionRequest. That or SomethingMessage? `TaskExecutionMessage`.

(Sorry to be pick about names, but I think it's important here.)

Reply:
Oh and I think the class should live under airflow.executors instead of airflow.models -- it's not stored in the DB which _most_ of in that module are.

Reply:
TaskExecutionRequest sounds good to me too.

Reply:
`TaskExecutionRequest` +1